crates/kubelet/src/kubelet.rs

```diff
@@ -49,6 +55,97 @@ pub enum Phase {
 pub struct Status {
     pub phase: Phase,
     pub message: Option<String>,
+    pub container_statuses: Vec<ContainerStatus>,
+}
+
+/// A struct that contains all necessary information for a waiting status, meant
+/// to be used with the [ContainerStatus] enum
+#[derive(Clone, Debug)]
+pub struct WaitingStatus {
+    /// The timestamp of when this status was reported
+    pub timestamp: DateTime<Utc>,
+    /// A human readable string describing the why it is in a waiting status
+    pub message: String,
+}
+
+/// A struct that contains all necessary information for a running status, meant
+/// to be used with the [ContainerStatus] enum
+#[derive(Clone, Debug)]
+pub struct RunningStatus {
+    /// The timestamp of when this status was reported
+    pub timestamp: DateTime<Utc>,
+}
+
+/// A struct that contains all necessary information for a terminated status,
+/// meant to be used with the [ContainerStatus] enum
+#[derive(Clone, Debug)]
+pub struct TerminatedStatus {
+    /// The timestamp of when this status was reported
+    pub timestamp: DateTime<Utc>,
+    /// A human readable string describing the why it is in a terminating status
+    pub message: String,
+    /// Should be set to true if the process exited with an error
+    pub failed: bool,
+}
+
+/// ContainerStatus is a simplified version of the Kubernetes container status
+/// for use in providers. It allows for simple creation of the current status of
+/// a "container" (a running wasm process) without worrying about a bunch of
+/// Options. Use the [ContainerStatus::to_kubernetes] method for converting it
+/// to a Kubernetes API status
+#[derive(Clone, Debug)]
+pub enum ContainerStatus {
+    Waiting(WaitingStatus),
+    Running(RunningStatus),
+    Terminated(TerminatedStatus),
+}
+
+impl ContainerStatus {
+    pub fn to_kubernetes(&self, pod_name: &str) -> KubeContainerStatus {
```

Comment:
If we need this to be a `String` anyway, we should just require a `String`. This prevents `String => &str => String` conversions.

Reply:
I think I’ll change this, but for discussion sake: What are your thoughts when it comes to convenience for API use though? Others who consume the crate might pass a literal or only have a borrow and it feels nicer to not force the to clone or use into string. Is there general guidance for this when making rust apis?

Reply:
I think I’ll change this, but for discussion sake: What are your thoughts when it comes to convenience for API use though? Others who consume the crate might pass a literal or only have a borrow and it feels nicer to not force the to clone or use into string. Is there general guidance for this when making rust apis?

Reply:
By taking &str here you're forcing a clone even if there doesn't need to be one. By taking a `String` here you're letting the caller know what you need and allowing them to give you it as efficiently as possible. With `&str` you'll have to clone every time even if the caller had an owned string they could give you.